src/wolfio.c

```diff
@@ -330,23 +329,43 @@ int EmbedReceiveFrom(WOLFSSL *ssl, char *buf, int sz, void *ctx)
     int recvd;
     int sd = dtlsCtx->rfd;
     int dtls_timeout = wolfSSL_dtls_get_current_timeout(ssl);
+    byte doDtlsTimeout;
     SOCKADDR_S peer;
     XSOCKLENT peerSz = sizeof(peer);
 
     WOLFSSL_ENTER("EmbedReceiveFrom()");
 
     /* Don't use ssl->options.handShakeDone since it is true even if
      * we are in the process of renegotiation */
-    if (ssl->options.handShakeState == HANDSHAKE_DONE)
+    doDtlsTimeout = ssl->options.handShakeState != HANDSHAKE_DONE;
+
+#ifdef WOLFSSL_DTLS13
+    if (ssl->options.dtls) {
+        doDtlsTimeout =
+            doDtlsTimeout || ssl->handshakeRtxFSM.rtxRecords != NULL ||
+            (ssl->dtls13FastTimeout && ssl->handshakeRtxFSM.ackRecords != NULL);
+    }
+#endif /* WOLFSSL_DTLS13 */
+
+    if (!doDtlsTimeout)
         dtls_timeout = 0;
 
     if (!wolfSSL_get_using_nonblock(ssl)) {
```

Comment:
Please also consider implementing a timeout mechanism for non-blocking sockets. You can leverage the `else if(IsSCR(ssl))` section to accomplish this. It currently only uses a `LowResTimer` but its better than nothing.

Reply:
I think that for non-blocking sockets, the application should take care of the dtls timeout. at least in this first beta version